File: test/Zorglub.Tests.CSharp/Time/Hemerology/YearNumberingTests.cs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2020 Narvalo.Org. All rights reserved.

namespace Zorglub.Time.Hemerology;

// TODO(code): dépassements arithmétiques.

using DataSet = YearNumberingDataSet;

public static class YearNumberingTests
{
    [Theory, MemberData(nameof(DataSet.CenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetCentury_Internal(CenturyInfo info) =>
        Assert.Equal(info.Century, YearNumbering.GetCentury(info.Year));

    [Theory, MemberData(nameof(DataSet.CenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetYearOfCentury_Internal(CenturyInfo info) =>
        Assert.Equal(info.YearOfCentury, YearNumbering.GetYearOfCentury(info.Year));

    //
    // Decades of the century
    //

    [Theory, MemberData(nameof(DataSet.DecadeOfCenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetDecadeOfCentury(DecadeOfCenturyInfo info)
    {
        var (y, century, decadeOfCentury, yearOfDecade) = info;
        // Act
        int decadeOfCenturyA = YearNumbering.GetDecadeOfCentury(
            y, out int centuryA, out int yearOfDecadeA);
        // Assert
        Assert.Equal(century, centuryA);
        Assert.Equal(decadeOfCentury, decadeOfCenturyA);
        Assert.Equal(yearOfDecade, yearOfDecadeA);
    }

    [Theory, MemberData(nameof(DataSet.IsoDecadeOfCenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetIsoDecadeOfCentury(DecadeOfCenturyInfo info)
    {
        var (y, century, decadeOfCentury, yearOfDecade) = info;
        // Act
        int decadeOfCenturyA = YearNumbering.GetIsoDecadeOfCentury(
            y, out int centuryA, out int yearOfDecadeA);
        // Assert
        Assert.Equal(century, centuryA);
        Assert.Equal(decadeOfCentury, decadeOfCenturyA);
        Assert.Equal(yearOfDecade, yearOfDecadeA);
    }

    //
    // Decades
    //

    [Theory, MemberData(nameof(DataSet.DecadeInfoData), MemberType = typeof(DataSet))]
    public static void GetDecade(DecadeInfo info)
    {
        var (y, decade, yearOfDecade) = info;
        // Act
        int decadeA = YearNumbering.GetDecade(y, out int yearOfDecadeA);
        // Assert
        Assert.Equal(decade, decadeA);
        Assert.Equal(yearOfDecade, yearOfDecadeA);
    }

    [Theory, MemberData(nameof(DataSet.DecadeInfoData), MemberType = typeof(DataSet))]
    public static void GetYearFromDecade(DecadeInfo info)
    {
        var (y, decade, yearOfDecade) = info;
        // Act
        int actual = YearNumbering.GetYearFromDecade(decade, yearOfDecade);
        // Assert
        Assert.Equal(y, actual);
    }

    [Theory, MemberData(nameof(DataSet.IsoDecadeInfoData), MemberType = typeof(DataSet))]
    public static void GetIsoDecade(DecadeInfo info)
    {
        var (y, decade, yearOfDecade) = info;
        // Act
        int decadeA = YearNumbering.GetIsoDecade(y, out int yearOfDecadeA);
        // Assert
        Assert.Equal(decade, decadeA);
        Assert.Equal(yearOfDecade, yearOfDecadeA);
    }

    [Theory, MemberData(nameof(DataSet.IsoDecadeInfoData), MemberType = typeof(DataSet))]
    public static void GetYearFromIsoDecade(DecadeInfo info)
    {
        var (y, decade, yearOfDecade) = info;
        // Act
        int actual = YearNumbering.GetYearFromIsoDecade(decade, yearOfDecade);
        // Assert
        Assert.Equal(y, actual);
    }

    //
    // Centuries
    //

    [Theory, MemberData(nameof(DataSet.CenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetCentury(CenturyInfo info)
    {
        var (y, century, yearOfCentury) = info;
        // Act
        int centuryA = YearNumbering.GetCentury(y, out int yearOfCenturyA);
        // Assert
        Assert.Equal(century, centuryA);
        Assert.Equal(yearOfCentury, yearOfCenturyA);
    }

    [Theory, MemberData(nameof(DataSet.CenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetYearFromCentury(CenturyInfo info)
    {
        var (y, century, yearOfCentury) = info;
        // Act
        int actual = YearNumbering.GetYearFromCentury(century, yearOfCentury);
        // Assert
        Assert.Equal(y, actual);
    }

    [Theory, MemberData(nameof(DataSet.IsoCenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetIsoCentury(CenturyInfo info)
    {
        var (y, century, yearOfCentury) = info;
        // Act
        int centuryA = YearNumbering.GetIsoCentury(y, out int yearOfCenturyA);
        // Assert
        Assert.Equal(century, centuryA);
        Assert.Equal(yearOfCentury, yearOfCenturyA);
    }

    [Theory, MemberData(nameof(DataSet.IsoCenturyInfoData), MemberType = typeof(DataSet))]
    public static void GetYearFromIsoCentury(CenturyInfo info)
    {
        var (y, century, yearOfCentury) = info;
        // Act
        int actual = YearNumbering.GetYearFromIsoCentury(century, yearOfCentury);
        // Assert
        Assert.Equal(y, actual);
    }

    //
    // Millennia
    //

    [Theory, MemberData(nameof(DataSet.MillenniumInfoData), MemberType = typeof(DataSet))]
    public static void GetMillennium(MillenniumInfo info)
    {
        var (y, millennium, yearOfMillennium) = info;
        // Act
        int millenniumA = YearNumbering.GetMillennium(y, out int yearOfMillenniumA);
        // Assert
        Assert.Equal(millennium, millenniumA);
        Assert.Equal(yearOfMillennium, yearOfMillenniumA);
    }

    [Theory, MemberData(nameof(DataSet.MillenniumInfoData), MemberType = typeof(DataSet))]
    public static void GetYearFromMillennium(MillenniumInfo info)
    {
        var (y, millennium, yearOfMillennium) = info;
        // Act
        int actual = YearNumbering.GetYearFromMillennium(millennium, yearOfMillennium);
        // Assert
        Assert.Equal(y, actual);
    }

    [Theory, MemberData(nameof(DataSet.IsoMillenniumInfoData), MemberType = typeof(DataSet))]
    public static void GetIsoMillennium(MillenniumInfo info)
    {
        var (y, millennium, yearOfMillennium) = info;
        // Act
        int millenniumA = YearNumbering.GetIsoMillennium(y, out int yearOfMillenniumA);
        // Assert
        Assert.Equal(millennium, millenniumA);
        Assert.Equal(yearOfMillennium, yearOfMillenniumA);
    }

    [Theory, MemberData(nameof(DataSet.IsoMillenniumInfoData), MemberType = typeof(DataSet))]
    public static void GetYearFromIsoMillennium(MillenniumInfo info)
    {
        var (y, millennium, yearOfMillennium) = info;
        // Act
        int actual = YearNumbering.GetYearFromIsoMillennium(millennium, yearOfMillennium);
        // Assert
        Assert.Equal(y, actual);
    }
}

File: test/Zorglub.Tests/Tests/Hemerology/YearNumberingTests.fs
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) Tran Ngoc Bich. All rights reserved.

module Zorglub.Tests.Hemerology.YearNumberingTests

open Zorglub.Testing
open Zorglub.Testing.Data

open Zorglub.Time.Hemerology

open Xunit

// TODO(code): dépassements arithmétiques.

let decadeOfCenturyInfoData = YearNumberingDataSet.DecadeOfCenturyInfoData
let isoDecadeOfCenturyInfoData = YearNumberingDataSet.IsoDecadeOfCenturyInfoData
let decadeInfoData = YearNumberingDataSet.DecadeInfoData
let isoDecadeInfoData = YearNumberingDataSet.IsoDecadeInfoData
let centuryInfoData = YearNumberingDataSet.CenturyInfoData
let isoCenturyInfoData = YearNumberingDataSet.IsoCenturyInfoData
let millenniumInfoData = YearNumberingDataSet.MillenniumInfoData
let isoMillenniumInfoData = YearNumberingDataSet.IsoMillenniumInfoData

[<Theory; MemberData(nameof(centuryInfoData))>]
let ``GetCentury() plain`` (info: CenturyInfo) =
    YearNumbering.GetCentury(info.Year) === info.Century

[<Theory; MemberData(nameof(centuryInfoData))>]
let ``GetYearOfCentury()`` (info: CenturyInfo) =
    YearNumbering.GetYearOfCentury(info.Year) === int(info.YearOfCentury)

//
// Decades of the century
//

[<Theory; MemberData(nameof(decadeOfCenturyInfoData))>]
let ``GetDecadeOfCentury()`` (info: DecadeOfCenturyInfo) =
    let (decadeOfCentury, century, yearOfDecade) = YearNumbering.GetDecadeOfCentury(info.Year)

    century         === info.Century
    decadeOfCentury === int(info.DecadeOfCentury)
    yearOfDecade    === int(info.YearOfDecade)

[<Theory; MemberData(nameof(isoDecadeOfCenturyInfoData))>]
let ``GetIsoDecadeOfCentury()`` (info: DecadeOfCenturyInfo) =
    let (decadeOfCentury, century, yearOfDecade) = YearNumbering.GetIsoDecadeOfCentury(info.Year)

    century         === info.Century
    decadeOfCentury === int(info.DecadeOfCentury)
    yearOfDecade    === int(info.YearOfDecade)

//
// Decades
//

[<Theory; MemberData(nameof(decadeInfoData))>]
let ``GetDecade()`` (info: DecadeInfo) =
    let (decade, yearOfDecade) = YearNumbering.GetDecade(info.Year)

    decade       === info.Decade
    yearOfDecade === int(info.YearOfDecade)

[<Theory; MemberData(nameof(decadeInfoData))>]
let ``GetYearFromDecade()`` (info: DecadeInfo) =
    let (y, decade, yearOfDecade) = info.Deconstruct()

    YearNumbering.GetYearFromDecade(decade, int(yearOfDecade)) === y

[<Theory; MemberData(nameof(isoDecadeInfoData))>]
let ``GetIsoDecade()`` (info: DecadeInfo) =
    let (decade, yearOfDecade) = YearNumbering.GetIsoDecade(info.Year)

    decade       === info.Decade
    yearOfDecade === int(info.YearOfDecade)

[<Theory; MemberData(nameof(isoDecadeInfoData))>]
let ``GetYearFromIsoDecade()`` (info: DecadeInfo) =
    let (y, decade, yearOfDecade) = info.Deconstruct()

    YearNumbering.GetYearFromIsoDecade(decade, int(yearOfDecade)) === y

//
// Centuries
//

[<Theory; MemberData(nameof(centuryInfoData))>]
let ``GetCentury()`` (info: CenturyInfo) =
    let (century, yearOfCentury) = YearNumbering.GetCentury(info.Year)

    century       === info.Century
    yearOfCentury === int(info.YearOfCentury)

[<Theory; MemberData(nameof(centuryInfoData))>]
let ``GetYearFromCentury()`` (info: CenturyInfo) =
    let (y, century, yearOfCentury) = info.Deconstruct()

    YearNumbering.GetYearFromCentury(century, int(yearOfCentury)) === y

[<Theory; MemberData(nameof(isoCenturyInfoData))>]
let ``GetIsoCentury()`` (info: CenturyInfo) =
    let (century, yearOfCentury) = YearNumbering.GetIsoCentury(info.Year)

    century       === info.Century
    yearOfCentury === int(info.YearOfCentury)

[<Theory; MemberData(nameof(isoCenturyInfoData))>]
let ``GetYearFromIsoCentury()`` (info: CenturyInfo) =
    let (y, century, yearOfCentury) = info.Deconstruct()

    YearNumbering.GetYearFromIsoCentury(century, int(yearOfCentury)) === y

//
// Millennia
//

[<Theory; MemberData(nameof(millenniumInfoData))>]
let ``GetMillennium()`` (info: MillenniumInfo) =
    let (millennium, yearOfMillennium) = YearNumbering.GetMillennium(info.Year)

    millennium       === info.Millennium
    yearOfMillennium === int(info.YearOfMillennium)

[<Theory; MemberData(nameof(millenniumInfoData))>]
let ``GetYearFromMillennium()`` (info: MillenniumInfo) =
    let (y, millennium, yearOfMillennium) = info.Deconstruct()

    YearNumbering.GetYearFromMillennium(millennium, int(yearOfMillennium)) === y

[<Theory; MemberData(nameof(isoMillenniumInfoData))>]
let ``GetIsoMillennium()`` (info: MillenniumInfo) =
    let (millennium, yearOfMillennium) = YearNumbering.GetIsoMillennium(info.Year)

    millennium       === info.Millennium
    yearOfMillennium === int(info.YearOfMillennium)

[<Theory; MemberData(nameof(isoMillenniumInfoData))>]
let ``GetYearFromIsoMillennium()`` (info: MillenniumInfo) =
    let (y, millennium, yearOfMillennium) = info.Deconstruct()

    YearNumbering.GetYearFromIsoMillennium(millennium, int(yearOfMillennium)) === y
